Align JSON output with original library

fdionisi opened the pull request:
To be more consistent with the original library result, `pattern` field in `struct Match` is flattened and `enum MatchPattern` type lowercased and moved to `pattern` field through [serde's attributes](https://serde.rs/attributes.html),  while `sub_display` is a join of `match_sub` values with `", "` using [itertools::Itertools::join](https://docs.rs/itertools/0.7.8/itertools/trait.Itertools.html#method.join).

shssoichiro commented:
Thanks! This looks good and I can see the value in it, but it is technically a breaking change. I'm fine with releasing a 2.0, but I'd like to make sure any other breaking changes go out at the same time, so if there are any other changes you think need to be made, please let me know and I'll bundle those in with a 2.0 release. I think it also wouldn't hurt to have a test for the serialization.

fdionisi commented:
That makes sense – I was wondering, what do you think if I also change `guesses_log10` to be `f64`? And at that point, making CrackTimes fields `f64` as well?
Also, how would you approach a serialization test without entering `serde_json` domain?
Just serialize an Entropy and compare it to a mocked one?

shssoichiro commented:
> I was wondering, what do you think if I also change guesses_log10 to be f64? And at that point, making CrackTimes fields f64 as well?

I'm fairly certain these are always whole numbers in the Javascript library. Feel free to correct me if I'm wrong.

> Just serialize an Entropy and compare it to a mocked one?

Yes, that should work. We can also include `serde_json` as a dev-dependency if needed.

shssoichiro commented:
Cool, I stand corrected then. Those can be `f64` for consistency.

fdionisi commented:
Sorry for my delay.

I encountered a problem with the serialisation of `HashMap` having `char` as key – apparently `serde_json` doesn't support that.
That means that serialising `Entropy` for "P@ssword!23" will panic.

I opened a [PR](https://github.com/serde-rs/json/pull/453) to allow `char` to be `MapKey`s.
I will update you and this PR as soon as I have news.

fdionisi commented:
Here we are.

Updated types, bump `serde_json` version and added serialisation test.
Let me know if this is enough or if I should do something more!

fdionisi commented:
Hey @shssoichiro,

Just checking in if can we move this forward?

shssoichiro commented:
Sorry, this slipped off my radar! Yes, I will go ahead and merge this. Thanks!
